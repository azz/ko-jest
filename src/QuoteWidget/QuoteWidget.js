const ko = require("knockout");

module.exports = class QuoteWidget {
  constructor(orderTicketInvoker, securitySearch) {
    this._orderTicketInvoker = orderTicketInvoker;
    this._securitySearch = securitySearch;

    this.securityCode = ko.observable("IRE.ASX");
    this.bidPrice = ko.observable(1);
    this.lastPrice = ko.observable(2);
    this.askPrice = ko.observable(3);
    this.hasTradingPermission = ko.observable(true);
  }

  onSearch(vm, event) {
    this._securitySearch.search(
      event.target.value,
      this.onSearchResult.bind(this)
    );
  }

  onSearchResult(response) {
    if (response.success) {
      this.securityCode(response.code);
    }
  }

  onBuy() {
    this._orderTicketInvoker.buy({
      code: this.securityCode(),
      price: this.bidPrice()
    });
  }

  onSell() {
    this._orderTicketInvoker.sell({
      code: this.securityCode(),
      price: this.askPrice()
    });
  }
};
