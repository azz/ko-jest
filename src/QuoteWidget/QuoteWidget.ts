
import * as ko from "knockout";

export default class QuoteWidget {
  securityCode = ko.observable("IRE.ASX");
  bidPrice = ko.observable(1);
  lastPrice = ko.observable(2);
  askPrice = ko.observable(3);
  hasTradingPermission = ko.observable(true);

  constructor(
    private _orderTicketInvoker,
    private _securitySearch,
    private _datePicker
  ) {}

  onSearch(vm: this, event) {
    this._securitySearch.search(
      event.target.value,
      this.onSearchResult.bind(this)
    );
  }

  onSearchResult(response) {
    if (response.success) {
      this.securityCode(response.code);
    }
  }

  onBuy() {
    this._orderTicketInvoker.buy({
      code: this.securityCode(),
      price: this.bidPrice()
    });
  }

  onSell() {
    this._orderTicketInvoker.sell({
      code: this.securityCode(),
      price: this.askPrice()
    });
  }
};
