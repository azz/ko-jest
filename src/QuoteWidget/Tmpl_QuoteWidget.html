<header>
    <input
        class="-security-input"
        data-bind="value: securityCode, event: { keyup: onSearch }"
    />
</header>
<main>
    <section>
        Bid Price: <span data-bind="text: bidPrice"></span>
    </section>
    <section>
        Last Price: <span data-bind="text: lastPrice"></span>
    </section>
    <section>
        Ask Price: <span data-bind="text: askPrice"></span>
    </section>

    <!-- ko if: hasTradingPermission() -->
    <button class="-buy-button" data-bind="click: onBuy">Buy</button>
    <button class="-sell-button" data-bind="click: onSell">Sell</button>
    <!-- /ko -->
</main>
