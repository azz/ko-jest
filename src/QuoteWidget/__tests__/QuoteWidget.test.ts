import {
  getTemplate,
  renderTemplateToDom,
  renderTemplateToRaw
} from "../../utils";
import * as ko from "knockout";
import * as path from "path";
import QuoteWidget from "../QuoteWidget";

const templatePath = path.resolve(__dirname, "..", "Tmpl_QuoteWidget.html");

describe("Quote Widget", () => {
  let script, widget, orderTicketInvoker, securitySearch, datePicker;

  const snapshot = () =>
    expect(
      renderTemplateToRaw(getTemplate(templatePath), widget)
    ).toMatchSnapshot();

  beforeEach(() => {
    // Mock a template
    // script = document.createElement('script');
    // script.type = "text/html";
    // script.id = "Tmpl_DatePicker";
    // script.innerHTML = `
    //   <div data-bind="text: currentDate></div>
    // `;
    // document.body.appendChild(script);


    orderTicketInvoker = {
      buy: jest.fn(),
      sell: jest.fn()
    };
    securitySearch = {
      search: jest.fn()
    };
    datePicker = {
      data: {
        currentDate: ko.observable(new Date())
      }
    }
    widget = new QuoteWidget(orderTicketInvoker, securitySearch, datePicker);
  });

  it("renders successfully", () => {
    snapshot();
  });

  it("renders altered last price", () => {
    widget.lastPrice(100);
    snapshot();
  });

  describe("buy/sell buttons", () => {
    it("are hidden if there's no trading permissions", () => {
      widget.hasTradingPermission(false);
      snapshot();
    });

    it("calls orderTicketInvoker.buy with bid price when the buy button is clicked", () => {
      widget.bidPrice(888);
      const { dom } = renderTemplateToDom(getTemplate(templatePath), widget);
      dom.querySelector(".-buy-button").click();
      expect(orderTicketInvoker.buy).toHaveBeenCalledWith({
        code: "IRE.ASX",
        price: 888
      });
    });

    it("calls orderTicketInvoker.sell with ask price when the sell button is clicked", () => {
      widget.askPrice(777);
      const { dom } = renderTemplateToDom(getTemplate(templatePath), widget);
      dom.querySelector(".-sell-button").click();
      expect(orderTicketInvoker.sell).toHaveBeenCalledWith({
        code: "IRE.ASX",
        price: 777
      });
    });
  });

  describe("search", () => {
    it("calls securitySearch.search with entered text on key up", () => {
      const { dom, window } = renderTemplateToDom(
        getTemplate(templatePath),
        widget
      );
      const input = dom.querySelector(".-security-input");
      expect(input).not.toBeNull();
      input.value = "BHP";
      input.dispatchEvent(new window.Event("keyup"));
      expect(securitySearch.search).toHaveBeenCalledWith(
        "BHP",
        expect.any(Function)
      );
    });

    it("changes value of search box when a successful result comes back", () => {
      const { dom } = renderTemplateToDom(getTemplate(templatePath), widget);
      securitySearch.search = jest.fn((text, cb) =>
        cb({
          success: true,
          code: text + ".ASX"
        })
      );
      widget.onSearch(null, { target: { value: "RIO" } });
      expect(dom.querySelector(".-security-input").value).toBe("RIO.ASX");
    });
  });
});
