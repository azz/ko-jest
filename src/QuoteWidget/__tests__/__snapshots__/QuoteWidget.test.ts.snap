// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quote Widget buy/sell buttons are hidden if there's no trading permissions 1`] = `
<div
  id="template"
>
  <header>
    <input
      class="-security-input"
    />
  </header>
  <main>
    <section>
      Blurst Price: 
      <span>
        1
      </span>
    </section>
    <section>
      Last Price: 
      <span>
        2
      </span>
    </section>
    <section>
      Ask Price: 
      <span>
        3
      </span>
    </section>
    <!-- /ko -->
  </main>
</div>
------------HTML PREVIEW---------------
<div
  id="template"
>
  <header>
    <input
      class="-security-input"
    />
  </header>
  <main>
    <section>
      Blurst Price: 
      <span>
        1
      </span>
    </section>
    <section>
      Last Price: 
      <span>
        2
      </span>
    </section>
    <section>
      Ask Price: 
      <span>
        3
      </span>
    </section>
    <!-- /ko -->
  </main>
</div>
`;

exports[`Quote Widget renders altered last price 1`] = `
<div
  id="template"
>
  <header>
    <input
      class="-security-input"
    />
  </header>
  <main>
    <section>
      Blurst Price: 
      <span>
        1
      </span>
    </section>
    <section>
      Last Price: 
      <span>
        100
      </span>
    </section>
    <section>
      Ask Price: 
      <span>
        3
      </span>
    </section>
    <button
      class="-buy-button"
    >
      Buy
    </button>
    <button
      class="-sell-button"
    >
      Sell
    </button>
  </main>
</div>
------------HTML PREVIEW---------------
<div
  id="template"
>
  <header>
    <input
      class="-security-input"
    />
  </header>
  <main>
    <section>
      Blurst Price: 
      <span>
        1
      </span>
    </section>
    <section>
      Last Price: 
      <span>
        100
      </span>
    </section>
    <section>
      Ask Price: 
      <span>
        3
      </span>
    </section>
    <button
      class="-buy-button"
    >
      Buy
    </button>
    <button
      class="-sell-button"
    >
      Sell
    </button>
  </main>
</div>
`;

exports[`Quote Widget renders successfully 1`] = `
<div
  id="template"
>
  <header>
    <input
      class="-security-input"
    />
  </header>
  <main>
    <section>
      Blurst Price: 
      <span>
        1
      </span>
    </section>
    <section>
      Last Price: 
      <span>
        2
      </span>
    </section>
    <section>
      Ask Price: 
      <span>
        3
      </span>
    </section>
    <button
      class="-buy-button"
    >
      Buy
    </button>
    <button
      class="-sell-button"
    >
      Sell
    </button>
  </main>
</div>
------------HTML PREVIEW---------------
<div
  id="template"
>
  <header>
    <input
      class="-security-input"
    />
  </header>
  <main>
    <section>
      Blurst Price: 
      <span>
        1
      </span>
    </section>
    <section>
      Last Price: 
      <span>
        2
      </span>
    </section>
    <section>
      Ask Price: 
      <span>
        3
      </span>
    </section>
    <button
      class="-buy-button"
    >
      Buy
    </button>
    <button
      class="-sell-button"
    >
      Sell
    </button>
  </main>
</div>
`;
